pkg/virt-launcher/virtwrap/api/converter.go

```diff
@@ -816,7 +816,21 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
 	if vmi.Spec.Domain.Memory != nil && vmi.Spec.Domain.Memory.Hugepages != nil {
 		domain.Spec.MemoryBacking = &MemoryBacking{
 			HugePages: &HugePages{},
+			// Workaround for the issue: https://github.com/kubevirt/kubevirt/issues/3781
+			Source: &MemoryBackingSource{Type: "memfd"},
+		}
+		// Set NUMA to use memfd
+		domain.Spec.CPU.NUMA = &NUMA{
+			Cells: []NUMACell{
+				{
+					ID:     "0",
+					CPUs:   fmt.Sprintf("0-%d", domain.Spec.VCPU.CPUs-1),
+					Memory: fmt.Sprintf("%d", getVirtualMemory(vmi).ScaledValue(resource.Mega)),
```

Comment:
This is still wrong. I need to assign correctly the unit

Reply:
I tried to modify in this way:
```diff
diff --git a/pkg/virt-launcher/virtwrap/api/converter.go b/pkg/virt-launcher/virtwrap/api/converter.go
index 270b27b41..6f3328080 100644
--- a/pkg/virt-launcher/virtwrap/api/converter.go
+++ b/pkg/virt-launcher/virtwrap/api/converter.go
@@ -825,8 +825,8 @@ func Convert_v1_VirtualMachine_To_api_Domain(vmi *v1.VirtualMachineInstance, dom
                                {
                                        ID:     "0",
                                        CPUs:   fmt.Sprintf("0-%d", domain.Spec.VCPU.CPUs-1),
-                                       Memory: fmt.Sprintf("%d", getVirtualMemory(vmi).ScaledValue(resource.Mega)),
-                                       Unit:   "MiB",
+                                       Memory: fmt.Sprintf("%d", getVirtualMemory(vmi).Value()/int64(1024)),
+                                       Unit:   "KiB",
                                },
                        },
                }
```
but the GiB test is still failing with:
```
------------------------------
• Failure [13.177 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1672]hugepages-1Gi [It]
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43

        Expected
            <int64>: 65011712
        to equal
            <int64>: 67108864

```

Reply:
It's getting closer. Let me pull your PR and run it locally.

Reply:
My suspicion is that the test is failing because the system is not totally clean up :/.
I just tried to run separately the tests:
Test 2Mi:
```bash
$ FUNC_TEST_ARGS='-ginkgo.focus=hugepages-2Mi' make functest
hack/dockerized "hack/build-func-tests.sh"
go version go1.12.8 linux/amd64

go version go1.12.8 linux/amd64
hack/functests.sh
Running Suite: Tests Suite
==========================
Random Seed: 1598515325
Will run 1 of 577 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
• [SLOW TEST:13.076 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1671]hugepages-2Mi
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Waiting for namespace kubevirt-test-default to be removed, this can take a while ...
Waiting for namespace kubevirt-test-alternative to be removed, this can take a while ...
Waiting for namespace kubevirt-test-operator to be removed, this can take a while ...

Ran 1 of 577 Specs in 25.413 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 576 Skipped
PASS

```
Test 1Gi:
```bash
$ FUNC_TEST_ARGS='-ginkgo.focus=hugepages-1Gi' make functest
hack/dockerized "hack/build-func-tests.sh"
go version go1.12.8 linux/amd64

go version go1.12.8 linux/amd64
hack/functests.sh
Running Suite: Tests Suite
==========================
Random Seed: 1598515143
Will run 2 of 577 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
S [SKIPPING] [4.607 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1672]hugepages-1Gi [It]
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43

        No node with hugepages hugepages-1Gi capacity

        /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:977
------------------------------
• [SLOW TEST:12.022 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1672]hugepages-1Gi
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Waiting for namespace kubevirt-test-default to be removed, this can take a while ...
Waiting for namespace kubevirt-test-alternative to be removed, this can take a while ...
Waiting for namespace kubevirt-test-operator to be removed, this can take a while ...

Ran 1 of 577 Specs in 48.889 seconds
SUCCESS! -- 1 Passed | 0 Failed | 0 Pending | 576 Skipped
PASS

```

All together:
```bash
FUNC_TEST_ARGS='-ginkgo.focus=hugepages' make functest
hack/dockerized "hack/build-func-tests.sh"
go version go1.12.8 linux/amd64

go version go1.12.8 linux/amd64
hack/functests.sh
Running Suite: Tests Suite
==========================
Random Seed: 1598515433
Will run 4 of 577 specs

SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
------------------------------
• [SLOW TEST:11.003 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1671]hugepages-2Mi
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43
------------------------------
S [SKIPPING] [2.603 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1672]hugepages-1Gi [It]
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43

        No node with hugepages hugepages-1Gi capacity

        /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:977
------------------------------
• Failure [12.079 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      should consume hugepages 
      /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [test_id:1672]hugepages-1Gi [It]
        /root/go/src/kubevirt.io/kubevirt/vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43

        Expected
            <int64>: 65011712
        to equal
            <int64>: 67108864

        /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:965
------------------------------
STEP: Starting a VM
{"component":"tests","kind":"VirtualMachineInstance","level":"info","msg":"Event(v1.ObjectReference{Kind:\"VirtualMachineInstance\", Namespace:\"kubevirt-test-default\", Name:\"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp\", UID:\"bf7de901-b744-492a-bda0-0d058ddc6466\", APIVersion:\"kubevirt.io/v1alpha3\", ResourceVersion:\"20095\", FieldPath:\"\"}): type: 'Normal' reason: 'SuccessfulCreate' Created virtual machine pod virt-launcher-testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxj7q5f","name":"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp","namespace":"kubevirt-test-default","pos":"utils.go:327","timestamp":"2020-08-27T08:04:13.945889Z","uid":"bf7de901-b744-492a-bda0-0d058ddc6466"}
{"component":"tests","kind":"VirtualMachineInstance","level":"info","msg":"Event(v1.ObjectReference{Kind:\"VirtualMachineInstance\", Namespace:\"kubevirt-test-default\", Name:\"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp\", UID:\"bf7de901-b744-492a-bda0-0d058ddc6466\", APIVersion:\"kubevirt.io/v1alpha3\", ResourceVersion:\"20155\", FieldPath:\"\"}): type: 'Normal' reason: 'Created' VirtualMachineInstance defined.","name":"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp","namespace":"kubevirt-test-default","pos":"utils.go:327","timestamp":"2020-08-27T08:04:20.091882Z","uid":"bf7de901-b744-492a-bda0-0d058ddc6466"}
{"component":"tests","kind":"VirtualMachineInstance","level":"info","msg":"Event(v1.ObjectReference{Kind:\"VirtualMachineInstance\", Namespace:\"kubevirt-test-default\", Name:\"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp\", UID:\"bf7de901-b744-492a-bda0-0d058ddc6466\", APIVersion:\"kubevirt.io/v1alpha3\", ResourceVersion:\"20160\", FieldPath:\"\"}): type: 'Normal' reason: 'Started' VirtualMachineInstance started.","name":"testvmiv4c8g97959b4fx9nvktcxvxjzpzj89gnnp6mxzd5fclx5gwp","namespace":"kubevirt-test-default","pos":"utils.go:327","timestamp":"2020-08-27T08:04:20.134800Z","uid":"bf7de901-b744-492a-bda0-0d058ddc6466"}
STEP: Checking that the VM memory equals to a number of consumed hugepages
failed to dump entity named [sriovnetworknodestates]: the server could not find the requested resource
failed to dump entity named [sriovnetworknodepolicies]: the server could not find the requested resource
failed to dump entity named [sriovnetworks]: the server could not find the requested resource
failed to dump entity named [sriovoperatorconfigs]: the server could not find the requested resource
failed to dump entity named [network-attachment-definitions]: the server could not find the requested resource
• [SLOW TEST:5.635 seconds]
Configurations
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:56
  VirtualMachineInstance definition
  /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:172
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:912
      with unsupported page size
      /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:1016
        [test_id:1673]should failed to schedule the pod
        /root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:1017
------------------------------
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Waiting for namespace kubevirt-test-default to be removed, this can take a while ...
Waiting for namespace kubevirt-test-alternative to be removed, this can take a while ...
Waiting for namespace kubevirt-test-operator to be removed, this can take a while ...


Summarizing 1 Failure:

[Fail] Configurations VirtualMachineInstance definition [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages should consume hugepages  [It] [test_id:1672]hugepages-1Gi 
/root/go/src/kubevirt.io/kubevirt/tests/vmi_configuration_test.go:965

Ran 3 of 577 Specs in 56.218 seconds
FAIL! -- 2 Passed | 1 Failed | 0 Pending | 574 Skipped
--- FAIL: TestTests (56.22s)
FAIL
make: *** [Makefile:69: functest] Error 1

```

Reply:
Just to confirmed that I added an ugly sleep in the code:
```diff 
git diff
diff --git a/tests/vmi_configuration_test.go b/tests/vmi_configuration_test.go
index d3735596a..4263b8c59 100644
--- a/tests/vmi_configuration_test.go
+++ b/tests/vmi_configuration_test.go
@@ -999,7 +999,7 @@ var _ = Describe("Configurations", func() {
                                        guestMemReq := resource.MustParse(guestMemory)
                                        hugepagesVmi.Spec.Domain.Memory.Guest = &guestMemReq
                                }
-
+                               time.Sleep(30 * time.Second)
                                By("Starting a VM")
                                _, err = virtClient.VirtualMachineInstance(tests.NamespaceTestDefault).Create(hugepagesVmi)
                                Expect(err).ToNot(HaveOccurred())
```
and the tests are now passing.

I need some feedback from you, what is worthy to implement. The proper way should be to loop until the free_pages==nr_pages  and rserv_page = 0 before starting the next testcase. It requiers of course some additional functions